### PR TITLE
[FLINK-4067] [runtime] Add savepoint headers

### PR DESCRIPTION
**Pull request by uce**
Savepoints were previously persisted without any meta data using default Java serialization of a `CompletedCheckpoint`. This PR introduces a savepoint class with version-specific serializers and stores savepoints with meta data.

Savepoints expose a version number and a `Collection<TaskState>` for savepoint restore.

Currently, there ~~are two savepoint versions~~ is only one savepoint version:

~~- `SavepointV0` (Flink 1.0): This is a wrapper around the Flink 1.0 `CompletedCheckpoint`. Because this class changed since Flink 1.0 I had to re-introduce it with this PR in order to allow restoring from Flink 1.0 savepoints (otherwise the old `StateForTask` state can not be deserialized via default Java serialization). Therefore, I've renamed the Flink 1.1 `CompletedCheckpoint` to `Checkpoint`. I've annotated the re-introduced classes with `@Deprecated`. We can remove those with Flink 1.2.~~
- `SavepointV0` (Flink 1.1): This is the current savepoint version, which holds a reference to the `Checkpoint` task state collection, but is serialized with a custom serializater not relying on default Java serialization. Therefore, it should not happen again that we need to stick to certain classes in future Flink versions.

The savepoints are stored in `FsSavepointStore` with the following format:

```
MagicNumber SavepointVersion Savepoint
  - MagicNumber => int
  - SavepointVersion => int (returned by Savepoint#getVersion())
  - Savepoint => bytes (serialized via version-specific SavepointSerializer)
```

The header is minimal (magic number, version). All savepoint-specific meta data can be moved to the savepoint itself. This is also were we would have to add new meta data in future versions, allowing us to differentiate between different savepoint versions when we change the serialization stack etc.

~~I've tested both manually and via a unit test that it works to trigger a savepoint via Flink 1.0 and resume from it with Flink 1.1.~~ It's not possible to resume from a Flink 1.0 savepoint.

All savepoint related classes have been moved from `checkpoint` to a new sub package `checkpoint.savepoint`.

The main classes to look at to review this PR are `Savepoint` and subclasses, `SavepointSerializer` and subclasses, and `FsSavepointStore`. @tillrohrmann do you have time to do this?

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoint.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	private final Map<JobVertexID, TaskState> taskStates;
+
+	public Checkpoint(
+		JobID job,
+		long checkpointID,
+		long timestamp,
+		long completionTimestamp,
+		Map<JobVertexID, TaskState> taskStates) {
+
+		this.job = job;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoint.java]: `checkNotNull`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java [hunk 2]
```diff
@@ -41,13 +41,13 @@
...
-	 * Creates a {@link CompletedCheckpointStore} instance for a job.
+	 * Creates a {@link CheckpointStore} instance for a job.
 	 *
 	 * @param jobId           Job ID to recover checkpoints for
 	 * @param userClassLoader User code class loader of the job
-	 * @return {@link CompletedCheckpointStore} instance for the job
+	 * @return {@link CheckpointStore} instance for the job
 	 */
-	CompletedCheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
+	CheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java]: Maybe we could rename this method to `createCheckpointStore` if not public.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 3]
```diff
@@ -109,13 +109,13 @@ public boolean isDiscarded() {
 		return discarded;
 	}
 	
-	public CompletedCheckpoint toCompletedCheckpoint() {
+	public Checkpoint toCompletedCheckpoint() {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java]: Maybe we could rename this method to something `finalizeCheckpoint` or `completeCheckpoint` if it's not public.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointRecoveryFactory.java [hunk 4]
```diff
@@ -37,10 +37,10 @@ public void stop() {
 	}
 
 	@Override
-	public CompletedCheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
+	public CheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointRecoveryFactory.java]: Renaming?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	@Override
+	public <T extends Savepoint> String storeSavepoint(T savepoint) throws IOException {
+		Preconditions.checkNotNull(savepoint, "Savepoint");
+
+		Exception latestException = null;
+		Path path = null;
+		FSDataOutputStream fdos = null;
+
+		// Try to create a FS output stream
+		for (int attempt = 0; attempt < 10; attempt++) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: What's usually the reason why `fileSystem.create` can fail? Why did you choose 9 retries?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+		return path.toString();
+	}
+
+	@Override
+	public Savepoint loadSavepoint(String path) throws IOException {
+		Preconditions.checkNotNull(path, "Path");
+
+		try {
+			try (FSDataInputStream fdis = createFsInputStream(new Path(path))) {
+				try (DataInputStream dis = new DataInputStream(fdis)) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: Why do we have a triple nested try block here? Wouldn't it also work to write:

```
try (FSDataInputStream fdis = createFsInputStream(new Path(path));
    DataInputStream dis = new DataInputStream(fdis)) {
    ...
} catch (Throwable t) {
...
}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+						Savepoint savepoint = serializer.deserialize(fdis);
+						return savepoint;
+					} else {
+						throw new IllegalStateException("Unexpected magic number. This indicates " +
+								"that the specified file is not a proper savepoint or " +
+								"the file has been corrupted.");
+					}
+				}
+			}
+		} catch (Throwable t) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: Exceptions which are thrown due to a wrong path or because the input stream couldn't be opened will also trigger the savepoint version 0 loading. Maybe we could differentiate between unable to open file stream and wrong magic number or unable to read version.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+	/**
+	 * Shut downs the savepoint store.
+	 *
+	 * <p>Only necessary for implementations where the savepoint life-cycle is
+	 * bound to the cluster life-cycle.
+	 *
+	 * @throws Exception Failures during shut down are forwarded
+	 */
+	void shutDown() throws Exception;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: I think so far we've usually written shutdown without a capital letter.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+			return fileSystem.open(path);
+		} else {
+			throw new IllegalArgumentException("Invalid path '" + path.toUri() + "'.");
+		}
+	}
+
+	/**
+	 * Tries to load a Flink 1.0 savepoint, which was not stored with a header.
+	 *
+	 * <p>This can be removed for Flink 1.2.
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: Why can we remove this for Flink 1.2? Wouldn't that mean that Flink 1.2 can only restore savepoints from version 1.1 onwards? I think it should also be possible to restore savepoints from version 1.0 with version 1.2.

**Comment by StephanEwen** [comment]
test comment

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+ * Serializer for {@link SavepointV0} instances.
+ *
+ * <p>This format was introduced with Flink 1.0.0 and simply wraps a
+ * {@link CompletedCheckpoint} instance and uses plain Java serialization.
+ * That's why we have to stick to the exact same format, including classes,
+ * which are only part of our code base for backwards compatability.
+ */
+public class SavepointV0Serializer implements SavepointSerializer<SavepointV0> {
+
+	public static final SavepointV0Serializer INSTANCE = new SavepointV0Serializer();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java]: Maybe we should add a private constructor so that nobody can instantiate this class.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	public void serialize(SavepointV0 savepoint, OutputStream os) throws IOException {
+		throw new UnsupportedOperationException("This serializer can only be used to " +
+				"deserialize Flink 1.0 savepoints. This version of Flink should only " +
+				"serialize more recent savepoints.");
+	}
+
+	@Override
+	public SavepointV0 deserialize(InputStream is) throws IOException {
+		try (ObjectInputStream ois = new ObjectInputStream(is)) {
+			try {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java]: Isn't one try block enough?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+		long checkpointId = ThreadLocalRandom.current().nextLong();
+
+		// Add serializer
+		serializers.put(version, NewSavepointSerializer.INSTANCE);
+
+		TestSavepoint newSavepoint = new TestSavepoint(version, checkpointId);
+		String pathNewSavepoint = store.storeSavepoint(newSavepoint);
+		assertEquals(1, tmp.getRoot().listFiles().length);
+
+		// Savepoint v0
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java]: savepoint v1

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+		// Load
+		Savepoint loaded = store.loadSavepoint(pathNewSavepoint);
+		assertEquals(newSavepoint, loaded);
+
+		loaded = store.loadSavepoint(pathSavepoint);
+		assertEquals(savepoint, loaded);
+	}
+
+	/**
+	 * Tests that an exception during store cleans up any created files.
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java]: Shouldn't we check that only the failed savepoint is cleaned up?

**Comment by tillrohrmann** [comment]
Really good work @uce. The code is well structured and thoroughly tested. I had only some minor comments.

While testing the code with the streaming state machine job I stumbled across a problem, though. Recovering from a Flink 1.0 savepoint does not work if the job contains a `keyBy` operation. The reason is that we had a faulty murmur hash implementation in Flink 1.0 and due to its correction, the mapping of keys to sub tasks has changed. Consequently, the restored state no longer matches the assigned key spaces for each operator. This is the problematic [commit](https://github.com/apache/flink/commit/641a0d436c9b7a34ff33ceb370cf29962cac4dee).

Thus, this change is actually breaking our backwards compatibility with respect to savepoints. In order to solve the problem I see three possibilities:
- Revert the changes of this commit. But we don't know how the flawed murmur hash performs.
- Develop a tool which can repartition savepoints
- Don't support backwards compatibility between version 1.0 and 1.1

I think that option 3 is not doable given our backwards compatibility promise. Furthermore, option 2 is not really straight forward, if the user has a keyed stream where he uses the `Checkpointed` interface. Given that the release is upcoming, I think option 1 would be the best way to solve the problem.

**Comment by uce** [comment]
Thanks for the thorough review! I didn't catch that by my manual test. I would also be in favour of 1) for the same reasons.

**Comment by uce** [comment]
@greghogan You fixed the murmur hash in the mentioned commit. Do you think it's OK to revert back to the wrong version we had before? Did you actually observe bad behaviour of the wrong murmur hash variant?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoint.java (hunk 1 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoint.java]: Will do (this is an unrelated change though as `Checkpoint` is a renaming of the current `master`'s `CompletedCheckpoint`)

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java (hunk 2 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java]: Will do, the current name is not good

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java (hunk 3 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java]: Will do

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointRecoveryFactory.java (hunk 4 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointRecoveryFactory.java]: Yes, to be consistent with the `CompletedCheckpoint => Checkpoint` renaming

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java (hunk 5 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: Since it can be also a distributed FS like HDFS, it can fail for various reasons. The code here is moved from what we currently have in master (including number of retries etc.). I think initially from `FileSystemStateStorageHelper` to `FileSystemStateStore` to here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java (hunk 6 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: True :smile:

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java (hunk 9 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: True, that should be defined outside of the code and we will probably like to keep it backwards compatible as long as possible. I will remove all comments like this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java (hunk 7 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java]: Good point, especially with opening the stream etc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java (hunk 8 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: We have it as `shutDown` in other places as well, but it's true that there are many places with `shutdown`. Since the method name describes an action I would like to keep it as two words *shut down*.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java (hunk 10 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java]: Forgot that, will add the private constructor

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java (hunk 11 above)
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java]: Oh yes 😄

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java (hunk 12 above)
Comment (uce) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java]: True

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java (hunk 13 above)
Comment (uce) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java]: Misleading formulation in the comment, that's exactly what's tested.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java (hunk 8 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: A quick search in the code base revealed that we have 5 usages of `shutDown()` and 100+ usages of `shutdown()`. I don't want to be hairsplitting, but I think the current convention is to write it without a capital letter.

Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: But the convention for virtually all other methods is that method names are typically verbs or verb phrases. But I don't want to be hairsplitting. I'll rename it.

**Comment by uce** [comment]
Thanks for the review again. I've addressed all comments. If you have time, it would be great to have a look over the change in 157c4e1. I've ran all tests locally after reverting FLINK-3623 and everything passes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java (hunk 8 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: I think *shutdown* is a verb in fact.

Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: http://www.dict.cc/?s=shut+down vs. http://www.dict.cc/?s=shutdown, but I really don't care :smile: Let's just leave it as `shutdown`.

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java]: Hmm dict.leo.org says it differently: http://dict.leo.org/ende/index_de.html#/search=shutdown&searchLoc=0&resultOrder=basic&multiwordShowSingle=on&pos=0

But on the other hand there is also this site: http://notaverb.com/shutdown.

**Comment by uce** [comment]
If there are no objections, I would like to merge this.

**Comment by uce** [comment]
After talking with @tillrohrmann and @StephanEwen, we discovered some obvious cases where restoring from 1.0 savepoints won't work in 1.1, for example if window state is involved, which is serialized in a different format since 1.0. I think the same is true for RocksDB.

Should we by default throw an Exception when a user tries to restore a Flink 1.0 savepoint with 1.1? On the other hand, there  are cases where it would work as expected... but if it does not work the error message will not be very helpful.

**Comment by uce** [comment]
Magic number mismatch will fail with an error message stating that a likely cause for this is resuming from Flink 1.0 savepoints, which is not supported. I've removed the legacy classes and undid the renamings. `SavepointV1` (Flink 1.1) has been down-graded to `SavepointV0` and the previous `SavepointV0` (for Flink 1.0) has been removed.
